esrally/track/params.py

```diff
@@ -953,39 +953,54 @@ def create_readers(
     create_reader,
 ):
     logger = logging.getLogger(__name__)
-    # pre-initialize in order to assign parallel tasks for different corpora through assignment
-    readers = [None for corpus in corpora for _ in corpus.documents] * num_clients
+    readers = []
+    total_readers = 0
     # stagger which corpus each client starts with for better parallelism
-    for group, corpus in enumerate(corpora[(start_client_index + mod) % len(corpora)] for mod in range(len(corpora))):
-        for entry, docs in enumerate(corpus.documents):
+    reordered_corpora = corpora[start_client_index:] + corpora[:start_client_index]
+    for corpus in reordered_corpora:
+        readers.append(collections.deque())
+        for docs in corpus.documents:
             offset, num_docs, num_lines = bounds(
                 docs.number_of_documents, start_client_index, end_client_index, num_clients, docs.includes_action_and_meta_data
             )
-            if num_docs > 0:
-                target = f"{docs.target_index}/{docs.target_type}" if docs.target_index else "/"
-                if docs.target_data_stream:
-                    target = docs.target_data_stream
-                logger.debug(
-                    "Task-relative clients at index [%d-%d] will bulk index [%d] docs starting from line offset [%d] for [%s] "
-                    "from corpus [%s].",
-                    start_client_index,
-                    end_client_index,
-                    num_docs,
-                    offset,
-                    target,
-                    corpus.name,
-                )
-                readers[len(corpora) * entry + group] = create_reader(
-                    docs, offset, num_lines, num_docs, batch_size, bulk_size, id_conflicts, conflict_probability, on_conflict, recency
-                )
-            else:
+            if num_docs == 0:
                 logger.debug(
                     "Task-relative clients at index [%d-%d] skip [%s] (no documents to read).",
                     start_client_index,
                     end_client_index,
                     corpus.name,
                 )
-    return readers
+                continue
+
+            target = f"{docs.target_index}/{docs.target_type}" if docs.target_index else "/"
+            if docs.target_data_stream:
+                target = docs.target_data_stream
+            logger.debug(
+                "Task-relative clients at index [%d-%d] will bulk index [%d] docs starting from line offset [%d] for [%s] "
+                "from corpus [%s].",
+                start_client_index,
+                end_client_index,
+                num_docs,
+                offset,
+                target,
+                corpus.name,
+            )
+            reader = create_reader(
+                docs, offset, num_lines, num_docs, batch_size, bulk_size, id_conflicts, conflict_probability, on_conflict, recency
+            )
+            readers[-1].append(reader)
```

Comment:
How about we assign the dequeue to a variable? This might be easier to read?

Reply:
Done in [`6870764` (#1659)](https://github.com/elastic/rally/pull/1659/commits/6870764ad9453f5ec8e54265bcc243c6ee4fafa2)